### ISSUE TITLE
Docker image with ANTsR ready to go?

### ISSUE THREAD

**Issue by dorianps** [ANTsX/ANTsR#265]
I was wondering why we don't have a docker image with ANTsR installed in it. Has anybody thought about this? Is there any obstacle from putting anything in docker? In my limited experience  docker images can get the user up and running in a matter of minutes without any overhead in thinking OS related issues, dependencies, misconfigurations, etc.

I have not build an image myself yet, but it shouldn't be difficult. particularly with tools like neurodocker.

**Comment by cookpa** [comment]
I don't have a Docker yet but I made a Singularity recipe

https://github.com/cookpa/ANTsRSingularity

**Comment by stnava** [comment]
been around since summer  2017
https://github.com/stnava/ANTsRDocker

**Comment by stnava** [comment]
also:  https://rstudio.cloud/project/38492

which is out of date as well.

**Comment by dorianps** [comment]
> also: https://rstudio.cloud/project/38492
> 
> which is out of date as well.

Checked that few days ago, doesn't work, it complained something about ANTsRCore.

Is the ANTsR docker image somewhere (docker hub?) or does it need a build?

Short story, I am looking to perhaps build an image with LESYMAP/ANTsR/ANTsRCore/ITKR.

**Comment by stnava** [comment]
https://hub.docker.com/r/stnava/antsr/   again 2 years old

**Comment by stnava** [comment]
also - the cloud version does work, at least when i've tried it.    "not working" seems very unlikely since it's just a docker container.

**Comment by dorianps** [comment]
```
> library(ANTsR)
Error in library(ANTsR) : there is no package called ‘ANTsR’
> .libPaths('/cloud/project/RL/')
> library(ANTsR)
Loading required package: ANTsRCore
Error: package or namespace load failed for ‘ANTsRCore’:
 package ‘ANTsRCore’ was installed by an R version with different internals; it needs to be reinstalled for use with this R version
Error: package ‘ANTsRCore’ could not be loaded
```

![image](https://user-images.githubusercontent.com/9083517/54717015-93651e80-4b2d-11e9-83a1-9654494cf962.png)

**Comment by stnava** [comment]
wrong R version relative to the source.    try R version 3.4.x

**Comment by stnava** [comment]
upper right corner.

**Comment by dorianps** [comment]
Got it, I managed to install devtools and LESYMAP as well, although example(lesymap) which normally takes 1 minute leads to an R crash.

In general it would be great to have a recent docker image and perhaps add docker hub tags every new github tag.

**Comment by stnava** [comment]
that would be nice.

brian

**Comment by stnava** [comment]
here is the latest Dockerfile :  https://github.com/stnava/ANTsRDocker   it is not pushed to hub - I have not found adequate wireless bandwidth to make a successful push .... it can be used in [https://mybinder.org](https://mybinder.org) though

**Comment by dorianps** [comment]
An update on docker efforts:

I have created an ANTsR docker image here:
https://hub.docker.com/r/dorianps/antsr/tags

I wanted to share a couple of thoughts/questions on the docker image.

### The docker image:
The image I build is based on rocker/rstudio and contains a fresh build of ANTsR using `devtools`. This guarantees we get in the build the latest changes (ANTsR, ANTsRCore, ITKR, ANTs), for example the latest bug fix by @cookpa . We also don't need to track the dependencies that are needed, we can just change dependencies in one of the packages and they are automatically reflected in the docker build. Looks like the ANTsRDocker by @stnava uses binary installs, so they will be behind with the latest fixes. In terms of size, the base image rocker/rstudio is much lighter than the one used by @stnava rocker/binder. For this, we end up with an antsr docker image with 2.7Gb in my solution as opposed to 7Gb at @stnava solution. And of course, I hope the idea is we provide the minimum environment required to run ANTsR and install additional packages. Since devtools is already in the environment, one can install other things on top (i.e., ANTsRNet etc.). This can be done on the fly once the container is running, or we can build extra images for the layers covering ANTsR, this is what I am doing for LINDA and LESYMAP.

### Keeping consistent tagging:
For normal software, people tag by version. But here we have all sorts of depndencies and makes little sense to tag by ANTsR version on an ongoing basis. Changes can happen in ANTsRCore or ITKR that are not reflected in ANTsR commits, and yet the ANTsR package has changed. So I propose we use current build dates for tagging docker images. Once R starts within the container, you should see the installed commit for each software, which can help us troubleshoot things if needed.
```
 package   * version date       lib source                          
 ANTsR       0.5.4   2019-10-29 [1] Github (ANTsX/ANTsR@e87caa4)    
 ANTsRCore   0.7.1   2019-10-29 [1] Github (ANTsX/ANTsRCore@9e17884)
 ITKR        0.5.1.2 2019-10-29 [1] Github (stnava/ITKR@2c73cf9)    

[1] /usr/local/lib/R/site-library
[2] /usr/local/lib/R/library
 package * version date       lib source
 LINDA   * 0.5.1   2019-10-30 [1] local 

[1] /usr/local/lib/R/site-library
[2] /usr/local/lib/R/library
```

### Autobuilds
Dockerhub offers automated image creation by tracking Github commits. This is not feasible for our style of work because commits can happen in ANTsR, ANTsRCore, or ITKR, and sometimes these commits are work in progress and the three components may be out of sync and fail the installation. I think the best would be to run a cron job and post docker images weekly or monthly. I can do that for now, but it might be better we keep a single common docker file in ANTsRDocker (or ANTsR itself) where everyone is able to perform pull requests to further refine the dockerfile, and then we use that dockerfile to somewhat be able build and push in docker hub. I can also make a single bash script with the docker file incorporated, which can run on any machine to build the docker image.

That's it for now. Happy to hear your thoughts. If anyone wants to try the docker image I built:
```
MYPASSWORD=mesi3 # set your password
MYDATA=/DATA/dorian # point to your data folder
LOCALPORT=8787

docker container run -e PASSWORD=$MYPASSWORD \
  -v $MYDATA:/home/rstudio/mydata \
  -p $LOCALPORT:8787 \
  --rm dorianps/antsr:latest
```
Then open your browser and go to http://localhost:8787 . I can confirm the image works on Windows docker as well, I was able to run a heavy LINDA automated lesion segmentation on Windows (although there is some clear mis-management of memory by Rscript which does not happen in R itself).

**Comment by dorianps** [comment]
@cookpa I saw something about it and Neurodocker provides some tools to get surgical environments, i.e., only antsRegistration. I have not experimented with that, but from what I understand there is not much room, except:
* Switch to an r-base environment, which could save 600-700mb. The r-base image starts with 700mb while the rstudio image starts with 1.3Gb.
* Avoid using or keeping `devtools`. The dependencies for devtools are not small, but I think it is useful to have everything in one place, devtools and rstudio, this way anyone can run the container and install anything else is needed. I would not favor keeping multiple versions of antsr containers (with and without rstudio), it would make things too complicated.


@muschellij2 
Perhaps  version tagging might make more sense for official releases. But with the current development page of these packages, I don't see how that would work.

**Comment by muratmaga** [comment]
I want to try this, but I am not familiar with the docker. I went as far as installing and running the docker instance (including setting up the password). However, it is not clear what the username should be for the Rstudio login (at 8787). It doesn’t accept my local username or root.


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by dorianps** [comment]
If you just copied the command above then: (username rstudio, pass mesi3)

**Comment by dorianps** [comment]
But that command would fail in Windows Powershell because of line breaks `\`. This is an example that works for Windows:
```
docker container run -e PASSWORD=mesi3 -v C:\Users\dp\Desktop\lindaTest\:/home/rstudio/mydata -p 8787:8787 --rm
dorianps/linda:latest
```

**Comment by muratmaga** [comment]
I actually ran it like this from powershell
docker run dorianps/antsr

I asked me to set up a password, I did. What I didn’t know the username, and now it works perfectly fine as far as I can tell.

Thanks for this. Looks very promising.
M

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by stnava** [comment]
one approach to versioning might be to build on ( and repair ) the ANTsVersions.csv ... the ANTsR version csv could also include the ANTsRCore and ITKR core versions that were used to build.   not perfect but perhaps better than current state.

**Comment by dorianps** [comment]
From my understanding:

The problem is not where to keep versions but that commits are made without
bumping versions. If ANTsRVersions.csv needs to be maintained with version
bumps after each commit, then we could just bump version in DESCRIPTION
itself. If bumping versions after each commit (or pull request) is not
happening, I find the current git commit in master to be the most
informative information on the package version. When I need to go back in
time with ANTsR, I typically get specific commits from each package (ANTsR,
ANTsCore, ITK) say from 3 months ago so that all three were at that same
timepoint regardless of versions; this is why I think it makes sense to tag
ANTsR docker images by date.

P.s. An update: docker can do automated builds in the cloud for free, but
my building from source is taking too long for the cloud compilation
limits. Will see if I can do anything about it, or will use my local
computer for docker.

**Comment by stnava** [comment]
'versions' and commits are conceptually different things.   we dont want to unify the concept for several reasons.   primarily, it's against the definition / purpose of versioning.

**Comment by dorianps** [comment]
Here are some updates.

DockerHub would theoretically work, but the build is too slow and the process is killed after a couple of hours. We could split up the build in 3-4 different images, but I don't like to coordinate multiple image builds.

I have improved the local builds and created some tools for everyone's use. There are in:
https://github.com/dorianps/docker/tree/master/antsr

`Builds.antsr.*` scripts will build the container without any user input in 1-2 hours. While  `Run.antsr.*` scripts will start the container and the browser so the user does not need to deal with commands and parameters.

Both the above is available for Linux and Windows. Windows users can just double click the .bat images and will have a fresh build or run of the container. This means Windows folks can do their work without any help from Linux access.

At the end of the above page there is a command you can run to see what commits are installed in the container.

I will probably add a Wiki page in LESYMAP or ANTsR to explain more of this stuff in detail.

Enjoy.

**Comment by dorianps** [comment]
I have finished the scripts needed to build and run docker images. 
https://github.com/dorianps/docker
The idea was to build or run docker containers with a single basic command. Currently setting up a crontab in a local server to build and push monthly builds of ANTsR images. Anyone can build and push images on Dockerhub though, all is needed is to change the username inside the script.

Closing the issue as now I think is resolved.
